File: Data Structures & Algorithms/house-robber-ii/submission-2.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums)==1:
            return nums[0]
        def robs(noms):
            memo=[]
            def robh(ints):
                ans=0
                if len(ints) <= len(memo) and len(memo)>1:
                    return memo[len(ints)-1]
                if len(ints)==1:
                    ans=ints[0]
                if len(ints)==2:
                    ans=max(ints[0],ints[1])
                if len(ints)>2:
                    ans=max(robh(ints[:len(ints)-2])+ints[-1],robh(ints[:len(ints)-1]))
                if len(ints)==len(memo)+1:
                    memo.append(ans)
                return ans
            return robh(noms)
        return max(robs(nums[:len(nums)-1]),robs(nums[1:]))

File: Data Structures & Algorithms/maximum-subarray/submission-2.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int ans=nums[0];
        int before=nums[0];
        for (int i=1; i<nums.size();++i) {
            int curr=max(before+nums[i], nums[i]);
            ans=max(curr, ans);
            before=curr;
        }
        return ans;
    }
};
